lib/src/default_index/revset_engine.rs

```diff
@@ -829,6 +829,43 @@ impl<'index> EvaluationContext<'index> {
                     Ok(Box::new(EagerRevset { positions }))
                 }
             }
+            ResolvedExpression::ReachableWithin { sources, domain } => {
+                let mut sets = union_find::UnionFind::<IndexPosition>::new();
+
+                // Compute all reachable subgraphs.
+                let domain_revset = self.evaluate(domain)?;
+                let domain: HashSet<_> = domain_revset.positions().attach(index).collect();
+                for pos in &domain {
+                    for parent_pos in index.entry_by_pos(*pos).parent_positions() {
+                        if domain.contains(&parent_pos) {
+                            sets.union(*pos, parent_pos);
+                        }
+                    }
+                }
+
+                // Identify disjoint sets reachable from sources.
+                let roots: HashSet<_> = intersection_by(
```

Comment:
nit: "roots" in this context might be confusing. Other algorithms are DAG-based, whereas parent links in `UnionFind` are unordered.

Maybe it can be renamed to "representatives"?